src/PipelineNet/ChainsOfResponsibility/ResponsibilityChain.cs

```diff
@@ -88,7 +128,41 @@ public TReturn Execute(TParameter parameter)
                     // the "next" function is assigned to the finally function or a 
                     // default empty function.
                     if (index == MiddlewareTypes.Count)
-                        func = this._finallyFunc ?? ((p) => default(TReturn));
+                    {
+                        if (_finallyType != null)
+                        {
+                            finallyResolverResult = MiddlewareResolver.Resolve(_finallyType);
+
+                            if (finallyResolverResult == null || finallyResolverResult.Middleware == null)
+                            {
+                                throw new InvalidOperationException($"'{MiddlewareResolver.GetType()}' failed to resolve finally of type '{_finallyType}'.");
+                            }
+
+                            if (finallyResolverResult.IsDisposable && !(finallyResolverResult.Middleware is IDisposable))
+                            {
+#if NETSTANDARD2_1_OR_GREATER
+                                if (finallyResolverResult.Middleware is IAsyncDisposable)
+                                {
+                                    throw new InvalidOperationException($"'{finallyResolverResult.Middleware.GetType()}' type only implements IAsyncDisposable." +
+                                        " Use AsyncResponsibilityChain to execute the configured pipeline.");
+                                }
+#endif
+
+                                throw new InvalidOperationException($"'{finallyResolverResult.Middleware.GetType()}' type does not implement IDisposable.");
+                            }
+
+                            var @finally = (IFinally<TParameter, TReturn>)finallyResolverResult.Middleware;
+                            func = (p) => @finally.Finally(p);
+                        }
+                        else if (_finallyFunc != null)
+                        {
+                            func = _finallyFunc;
+                        }
+                        else
+                        {
+                            func = (p) => default(TReturn);
+                        }
+                    }
```

Comment:
It looks like we can extract a pattern for checking the _disposability_ of the resolved middlewares. Maybe something like this:


```c#
namespace PipelineNet.MiddlewareResolver
{
    internal static class MiddlewareResolverExtensions
    {
        public static void ValidateDisposableValues(this MiddlewareResolverResult resolverResult)
        {
            if (resolverResult.IsDisposable && !(resolverResult.Middleware is IDisposable))
            {
#if NETSTANDARD2_1_OR_GREATER
                if (resolverResult.Middleware is IAsyncDisposable)
                {
                    throw new InvalidOperationException($"'{resolverResult.Middleware.GetType()}' type only implements IAsyncDisposable." +
                        " Use the AsyncResponsibilityChain or AsyncPipeline.");
                }
#endif

                throw new InvalidOperationException($"'{resolverResult.Middleware.GetType()}' type does not implement IDisposable.");
            }
        }
    }
}
```

Then this code could be simplified with something like this:


```c#
        // new method just to resolve the finally function and the disposable middleware:
        private Func<TParameter, TReturn> ResolveFinallyFunction(ref IDisposable finallyHandler)
        {
            Func<TParameter, TReturn> func;
            if (_finallyType != null)
            {
                var finallyResolverResult = MiddlewareResolver.Resolve(_finallyType);

                if (finallyResolverResult == null || finallyResolverResult.Middleware == null)
                {
                    throw new InvalidOperationException($"'{MiddlewareResolver.GetType()}' failed to resolve finally of type '{_finallyType}'.");
                }

                finallyResolverResult.ValidateDisposableValues();

                var @finally = (IFinally<TParameter, TReturn>) finallyResolverResult.Middleware;
                finallyHandler = @finally as IDisposable;

                func = (p) => @finally.Finally(p);
            }
            else if (_finallyFunc != null)
            {
                func = _finallyFunc;
            }
            else
            {
                func = (p) => default(TReturn);
            }

            return func;
        }
```

Then the method would be much simpler:
```c#
Func<TParameter, TReturn> func = null;
func = (param) =>
{
    MiddlewareResolverResult resolverResult = null;
    IDisposable disposableFinallyMiddleware = null;
    try
    {
        var type = MiddlewareTypes[index];
        resolverResult = MiddlewareResolver.Resolve(type);
        var middleware = (IMiddleware<TParameter, TReturn>)resolverResult.Middleware;

        index++;
        // If the current instance of middleware is the last one in the list,
        // the "next" function is assigned to the finally function or a 
        // default empty function.
        if (index == MiddlewareTypes.Count)
        {
            func = ResolveFinallyFunction(ref disposableFinallyMiddleware);
        }

        resolverResult.ValidateDisposableValues();

        return middleware.Run(param, func);
    }
    finally
    {
        if (resolverResult != null && resolverResult.IsDisposable)
        {
            var middleware = resolverResult.Middleware;
            if (middleware != null)
            {
                if (middleware is IDisposable disposable)
                {
                    disposable.Dispose();
                }
            }
        }

        disposableFinallyMiddleware?.Dispose();
    }
};
```

Reply:
Oh, definitely, also for disposal itself.

Could you please ignore the code duplication herę for now though? I have a refactoring PR which I can open after all current ones are merged.